src/librustc/sir.rs

```diff
@@ -164,11 +216,47 @@ impl SirCx {
         self.builders.insert(builder, (bb, pos));
     }
 
-    /// Create a return void instruction in the SIR.
-    pub fn ret_void(&mut self, builder: *const Builder) {
+    /// Inserts a statement at the insertion point corresponding with `builder`.
+    pub fn emit(&mut self, builder: *const Builder, stmt: ykpack::Statement) {
         let (bb, idx) = self.builders[&builder];
         let sir_block = self.get_sir_block(bb);
-        let instr = ykpack::Statement::Return;
-        sir_block.stmts.insert(idx, instr);
+        sir_block.stmts.insert(idx, stmt);
+    }
+
+    /// Returns the function containing the insertion point of the supplied builder.
+    pub fn current_func(&mut self, builder: *const Builder) -> SirFuncIdx {
+        let (llbb, _) = self.builders[&builder];
+        self.llvm_blocks[&llbb].0
+    }
+
+    pub fn emit_assign(
+        &mut self,
+        builder: *const Builder,
+        result: SirLocalIdx,
+        rhs: ykpack::Rvalue,
+    ) {
+        self.emit(builder, ykpack::Statement::Assign(result.as_u32(), rhs));
+    }
+
+    pub fn emit_load(&mut self, builder: *const Builder, result: *const Value, arg: *const Value) {
+        // FIXME deal with volatile loads.
```

Comment:
Can we add an `assert` that will fail if we encounter such a load?

Reply:
We won't encounter one until we have added a call from the correct place in the LLVM backend. Somewhere in/around here:
https://github.com/softdevteam/ykrustc/blob/043bf15b2c4fd0b21c02f6fe3d82bb8f663faf75/src/librustc_codegen_llvm/builder.rs#L446

Reply:
That's not my point: when we do add it somewhere else, it's *much* better if every other point we need to touch explicitly crashes. Now I don't know if we can easily tell a volatile vs. a non-volatile load here, but if we can easily do so, let's change the `FIXME` to an `assert`.

Reply:
This raises a good point actually. Due to the way the LLVM codegen interacts with the SirCx, SIR's view of the world will be incomplete until we have identified and handled all of the places LLVM instructions are generated.

I think that's just going to have to be the way it is I'm afraid. There's no one 'insert an instruction' place that we can hook in to. They are scattered around the codegen.

Reply:
That still seems very orthogonal to this `FIXME` :)

Reply:
(sorry github wasn't updating, so my above response looks very out of context -- I thought I was adding additional context to my original response)

> it's much better if every other point we need to touch explicitly crashes

The compiler needs to be able to build itself in order for development to continue, so I don't think we can make it crash.

Reply:
We resolved this offline and decided we would identify all of the places where a Sir statement should be added and have them insert an "unknown" instruction.

This allows the compiler to build itself, but for us to know (at runtime) if Sir functionality is incomplete.

Reply:
This might be impractical/impossible, but if the "insert unknown instruction" thing (macro?) can also record where it was inserted (i.e. the source rust file), that might help us debugging in the future?

Reply:
I was going to put a string inside the instruction which identifies what kind of instruction wasn't handled. That's what we did before anyway.

https://github.com/softdevteam/yk/blob/7b4cfe28cf0148813ee8efd9b58c3c023321eafd/ykpack/src/types.rs#L175-L177

(The comment is wrong to talk about MIR, will fix).

Reply:
If a macro tells you the source file/line even better, but otherwise we can add it in manualy.

Reply:
There are macros in std to get those, yeah, but it's actually more convenient to have the name of the instruction in there.

Picture the scenario. You have a mir dump open in vim and see `$3 = unimplemented(file=blah.rs, line=456)`.

I'm more interested in what kind of instruction it is, and I'd have to go digging in the source to find out.

Thoughts?

Reply:
If you have `blah.rs:456` you can look it up in your editor in about 5 seconds. If it says "Add" you'll probably have 10 locations to grep for.

Of course, you can do both. `unknow_instr!(add)` can say "add" and `blah.rs:456`.

Reply:
I was going to say, let's do both, but I've just noticed that quite a lot of hook points are already inside macros. I'm worried this is going to get scary.

I'll give it a go, but reserve the right to wimp out :)

Reply:
I'm sure it'll be OK, but let's see.